### ISSUE TITLE
[DI] doctrine entity manager build order

### ISSUE THREAD

**Issue by axi** [symfony/symfony#48895]
### Symfony version(s) affected

5.4.17

### Description

After upgrading to symfony/dependency-injection 5.4.17, several users (including me) encountered the following exception when clearing cache:
```
In MissingMappingDriverImplementation.php line 11:
                                                                                                                       
  [Doctrine\ORM\Exception\MissingMappingDriverImplementation]                                                          
  It's a requirement to specify a Metadata Driver and pass it to Doctrine\ORM\Configuration::setMetadataDriverImpl().  

Exception trace:
  at /var/www/app/vendor/doctrine/orm/lib/Doctrine/ORM/Exception/MissingMappingDriverImplementation.php:11
 Doctrine\ORM\Exception\MissingMappingDriverImplementation::create() at /var/www/app/vendor/doctrine/orm/lib/Doctrine/ORM/EntityManager.php:162
 Doctrine\ORM\EntityManager->__construct() at /var/www/app/var/cache/dev/ContainerXpmlFIp/srcApp_KernelDevDebugContainer.php:4128
 ```
 Rolling back to 5.4.16 fix the exception.
 We started discussing this in the #48814 issue but it finaly doesn't look like the same problem.

------ Following is the first comment from #48814 by @julienfastre ----------------
 
I experience a problem which may be similar, with a complex configuration of `EntityManagerInterface`.

With symfony/dependency-injection on 5.4.17, the kernel "prepares" the EntityManager like this:

```php
    /**
     * Gets the public 'doctrine.orm.default_entity_manager' shared service.
     *
     * @return \Doctrine\ORM\EntityManager
     */
    protected function getDoctrine_Orm_DefaultEntityManagerService($lazyLoad = true)
    {
        $a = ($this->services['doctrine.dbal.default_connection'] ?? $this->getDoctrine_Dbal_DefaultConnectionService());

        if (isset($this->services['doctrine.orm.default_entity_manager'])) {
            return $this->services['doctrine.orm.default_entity_manager'];
        }
        $b = new \Doctrine\ORM\Configuration();

        $c = new \Doctrine\Persistence\Mapping\Driver\MappingDriverChain();

        $d = ($this->privates['doctrine.orm.default_annotation_metadata_driver'] ?? $this->getDoctrine_Orm_DefaultAnnotationMetadataDriverService());

        $c->addDriver($d, 'App\\Entity');
        $c->addDriver($d, 'Chill\\MainBundle\\Entity');
        $c->addDriver($d, 'Chill\\PersonBundle\\Entity');
        $c->addDriver($d, 'Chill\\CustomFieldsBundle\\Entity');
        $c->addDriver($d, 'Chill\\ActivityBundle\\Entity');
        $c->addDriver($d, 'Chill\\DocStoreBundle\\Entity');
        $c->addDriver($d, 'Chill\\DocGeneratorBundle\\Entity');
        $c->addDriver($d, 'Chill\\AsideActivityBundle\\Entity');
        $c->addDriver($d, 'Chill\\CalendarBundle\\Entity');
        $c->addDriver($d, 'Chill\\TaskBundle\\Entity');
        $c->addDriver($d, 'Chill\\ThirdPartyBundle\\Entity');
        $c->addDriver($d, 'Chill\\BudgetBundle\\Entity');

        // here, an exception is thrown because the configuration ($b) does not contains a mapping driver,
       // which is added later in the function
        $this->services['doctrine.orm.default_entity_manager'] = $instance = new \Doctrine\ORM\EntityManager($a, $b);

        (new \Doctrine\Bundle\DoctrineBundle\ManagerConfigurator([], []))->configure($instance);

        $b->setEntityNamespaces(['App' => 'App\\Entity', 'ChillMainBundle' => 'Chill\\MainBundle\\Entity', 'ChillPersonBundle' => 'Chill\\PersonBundle\\Entity', 'ChillCustomFieldsBundle' => 'Chill\\CustomFieldsBundle\\Entity', 'ChillActivityBundle' => 'Chill\\ActivityBundle\\Entity', 'ChillDocStoreBundle' => 'Chill\\DocStoreBundle\\Entity', 'ChillDocGeneratorBundle' => 'Chill\\DocGeneratorBundle\\Entity', 'ChillAsideActivityBundle' => 'Chill\\AsideActivityBundle\\Entity', 'ChillCalendarBundle' => 'Chill\\CalendarBundle\\Entity', 'ChillTaskBundle' => 'Chill\\TaskBundle\\Entity', 'ChillThirdPartyBundle' => 'Chill\\ThirdPartyBundle\\Entity', 'ChillBudgetBundle' => 'Chill\\BudgetBundle\\Entity']);
        $b->setMetadataCache(new \Symfony\Component\Cache\Adapter\ArrayAdapter());
        $b->setQueryCache(($this->privates['cache.doctrine.orm.default.query'] ?? $this->getCache_Doctrine_Orm_Default_QueryService()));
        $b->setResultCache(($this->privates['cache.doctrine.orm.default.result'] ?? $this->getCache_Doctrine_Orm_Default_ResultService()));

       /// here is added the missing metadata driver
        $b->setMetadataDriverImpl($c);

        $b->setProxyDir(($this->targetDir.''.'/doctrine/orm/Proxies'));
        $b->setProxyNamespace('Proxies');
        $b->setAutoGenerateProxyClasses(true);
        $b->setSchemaIgnoreClasses([]);
        $b->setClassMetadataFactoryName('Doctrine\\ORM\\Mapping\\ClassMetadataFactory');
        $b->setDefaultRepositoryClassName('Doctrine\\ORM\\EntityRepository');
        $b->setNamingStrategy(new \Doctrine\ORM\Mapping\DefaultNamingStrategy());
        $b->setQuoteStrategy(new \Doctrine\ORM\Mapping\DefaultQuoteStrategy());
        $b->setEntityListenerResolver(($this->services['doctrine.orm.default_entity_listener_resolver'] ?? $this->getDoctrine_Orm_DefaultEntityListenerResolverService()));
        $b->setRepositoryFactory( /* etc. etc. */)
        $b->addCustomHydrationMode('chill_flat_hierarchy_list', 'Chill\\MainBundle\\Doctrine\\ORM\\Hydration\\FlatHierarchyEntityHydrator');
        $b->addCustomStringFunction('GET_PERSON_ADDRESS_ADDRESS_ID', 'Chill\\PersonBundle\\Doctrine\\DQL\\AddressPart\\AddressPartAddressId');

        // ...
        // add a lot of other custom function there ...
       // ...

        return $instance;
    }
```

This cause a bug because the EntityManager instance is created too early:

```
In MissingMappingDriverImplementation.php line 11:
                                                                                                                       
  [Doctrine\ORM\Exception\MissingMappingDriverImplementation]                                                          
  It's a requirement to specify a Metadata Driver and pass it to Doctrine\ORM\Configuration::setMetadataDriverImpl().  
                                                                                                                       

Exception trace:
  at /var/www/app/vendor/doctrine/orm/lib/Doctrine/ORM/Exception/MissingMappingDriverImplementation.php:11
 Doctrine\ORM\Exception\MissingMappingDriverImplementation::create() at /var/www/app/vendor/doctrine/orm/lib/Doctrine/ORM/EntityManager.php:162
 Doctrine\ORM\EntityManager->__construct() at /var/www/app/var/cache/dev/ContainerXpmlFIp/srcApp_KernelDevDebugContainer.php:4128
```

If I move those two lines at the end of the method, just before the `return $instance;`, it does work:

```php
    protected function getDoctrine_Orm_DefaultEntityManagerService($lazyLoad = true)
    {
        // ...
        
        $this->services['doctrine.orm.default_entity_manager'] = $instance = new \Doctrine\ORM\EntityManager($a, $b);

        (new \Doctrine\Bundle\DoctrineBundle\ManagerConfigurator([], []))->configure($instance);

        return $instance;
    }
```

If I downgrade symfony/dependency-injection to 5.4.16 (`composer require symfony/dependency-injection 5.4.16`), the container is compiled in this manner, and the problem is fixed.

**EDIT 30/12/22** add comment in code block and correct version mismatch

_Originally posted by @julienfastre in https://github.com/symfony/symfony/issues/48814#issuecomment-1367672148_
      

### How to reproduce

I didn't succed to create a reproducer.

------ Following is a comment from #48814 by @HeinDR who tracked the bug down to a change from a commit ----------------

I also experience a similar problem with the doctrine configuration when I add an extra entity manager. After the update te symfony/dependency-injection 5.4.17 I get this error:
`It's a requirement to specify a Metadata Driver and pass it to Doctrine\ORM\Configuration::setMetadataDriverImpl().`
when I clear the cache.

I could trace the problem back to commit [58f2988](https://github.com/symfony/dependency-injection/commit/58f2988128d2d278280781db037677a32cf720db) where in the PhpDumper the `edge->isLazy()` check is removed in favor of checks further down the foreach.

```php
private function collectCircularReferences(string $sourceId, array $edges, array &$checkedNodes, array &$loops = [], array $path = [], bool $byConstructor = true): void
    {
        $path[$sourceId] = $byConstructor;
        $checkedNodes[$sourceId] = true;
        foreach ($edges as $edge) {
            $node = $edge->getDestNode();
            $id = $node->getId();
            if ($sourceId === $id || !$node->getValue() instanceof Definition || $edge->isWeak()) {
                continue;
            }
       //..
    }
```

When I add the `edge->isLazy()` check and manualy delete the cache the error is gone.

```php
private function collectCircularReferences(string $sourceId, array $edges, array &$checkedNodes, array &$loops = [], array $path = [], bool $byConstructor = true): void
    {
        $path[$sourceId] = $byConstructor;
        $checkedNodes[$sourceId] = true;
        foreach ($edges as $edge) {
            $node = $edge->getDestNode();
            $id = $node->getId();
            if ($sourceId === $id || !$node->getValue() instanceof Definition || $edge->isLazy() || $edge->isWeak()) {
                continue;
            }
       //..
    }
```

_Originally posted by @HeinDR in https://github.com/symfony/symfony/issues/48814#issuecomment-1371966269_

**Comment by axi** [comment]
I've got the same bug as @julienfastre
Here is a diff between 5.4.16 and 5.4.17 build so it might be easier to understand
https://gist.github.com/axi/7ffbb638b09f7f591099974c65bf135e/revisions

Also, this was introduced in https://github.com/symfony/symfony/pull/48791

**Comment by DavG** [comment]
Same bug here in one of my projects

```
 [KO]
Script cache:clear returned with error code 1
!!  
!!  In MissingMappingDriverImplementation.php line 11:
!!                                                                                 
!!    It's a requirement to specify a Metadata Driver and pass it to Doctrine\ORM  
!!    \Configuration::setMetadataDriverImpl().
```

I have of course this on my doctrine.yaml

```
doctrine:
    dbal:
        default_connection: default
        connections:
            default:
                driver: 'pdo_mysql'
                server_version: '5.7'
    orm:
        auto_generate_proxy_classes: true
        naming_strategy: doctrine.orm.naming_strategy.underscore
        auto_mapping: true
        mappings:
            App:
                is_bundle: false
                type: annotation
                dir: '%kernel.project_dir%/src/Entity'
                prefix: 'App\Entity'
                alias: App

```

Rolling back to 5.4.16 solves the issue for me.

@stof: yes I know, I edited my message. But I have the same conf on other projects and it's working with 5.4.17, so I was not sure.

**Comment by stof** [comment]
The exception talks about the mapping driver of the ORM. the config you show is the DBAL config, which is unrelated.

**Comment by xabbuh** [comment]
Can you create a small example application that allows to reproduce your issue?

**Comment by stefankleff** [comment]
I have the same issue after I've updated one of our services from di 6.1.8 to 6.1.9. A very similar service had no issues with the update. I already compared service definitions of the entity manager, configs, bundle loading order, etc. everything is exactly the same. I guess this is the reason why it's so hard to reproduce the error. 

Now that I've found this thread I'll dig deeper in the diff of the versions and try to find the change causing the behaviour.

**Comment by stefankleff** [comment]
I can confirm that https://github.com/symfony/symfony/pull/48791 introduced the bug. 
I'm not very familiar with the component and idk how to provide a simple example. But I would be open to have a screen sharing session or for further advice what I can provide to hunt down the bug.

**Comment by axi** [comment]
I'm working on a reproducer with no success until now... Maybe next week.

**Comment by stefankleff** [comment]
After about 10h I have a quite simple application causing the error:
https://github.com/stefankleff/symfony_di

The weird thing is, that it somehow depends on the amount of injections.
More Information is in the readme.

**Comment by axi** [comment]
Because I also spent hours, here is my reproducer on sf 5.4: https://github.com/axi/symfony_di_48895
But all the credits goes to @stefankleff because I was going around in circles and my reproducer use the same 2 listeners problem :-)

**Comment by stefankleff** [comment]
Some more insights. Maybe it helps you @axi or @xabbuh 

What happens is that from analyzeReferences() we get  contents in`$this->circularReferences` now. The array was empty before, but now it contains:
```
[
  "doctrine.orm.default_entity_manager" => array:1 [
    "doctrine.orm.default_entity_listener_resolver" => true
  ]
  "doctrine.orm.default_entity_listener_resolver" => array:2 [
    "App\EntityListener\My1EntityListener" => false
    "App\EntityListener\My2EntityListener" => true
  ]
  "App\EntityListener\My2EntityListener" => array:1 [
    "App\Service\MyService" => true
  ]
  "App\Service\MyService" => array:1 [
    "doctrine.orm.default_entity_manager" => true
  ]
  "App\EntityListener\My1EntityListener" => array:1 [
    "App\Service\MyService" => false
  ]
]
```
I can see that `addInlineService` for `doctrine.orm.default_entity_listener_resolver` is called while creating the code for `doctrine.orm.default_entity_manager`, which is then creating the piece of code with `new \Doctrine\ORM\EntityManager`.

**Comment by stefankleff** [comment]
I have a suspicion, but I still do not get the full picture. But it's either a bug or the solution to this bug:

https://github.com/symfony/symfony/blob/6.1/src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php#L478
```php
// we can now build the loop
$loop = null;
$loopByConstructor = $edge->isReferencedByConstructor() && !$edge->isLazy();
foreach ($fillPath as $k => $pathByConstructor) {
    if (null !== $loop) {
        $loop[] = $k;
        $loopByConstructor = $loopByConstructor && $pathByConstructor;
    } elseif ($k === $first) {
        $loop = [];
    }
}
$this->addCircularReferences($first, $loop, true);
```

The third argument of `addCircularReferences(string $sourceId, array $currentPath, bool $byConstructor)` in the last line is hard-coded. I don't see a reason why the value of `$loopByConstructor` is calculated, but in the end not used.

 If the argument is set to the variable it works for me.

**Comment by stefankleff** [comment]
It was introduced in this commit https://github.com/symfony/symfony/commit/1c3721e8adb5b1908086c9246888c9965405d10ebased based on this MR
https://github.com/symfony/symfony/pull/39058 in on Nov 12, 2020 for Symfony 5.1. 

@jderusse It seems like a bug to me

**Comment by Jean85** [comment]
https://github.com/getsentry/sentry-symfony/issues/689 seems another case where this issue is triggering a bug.

**Comment by nicolas-grekas** [comment]
Thanks for the work on this issue, especially for the reproducers!
I'm proposing to revert #48791 in #48962 as that's the PR that introduced the issue.

**Comment by stefankleff** [comment]
@nicolas-grekas That's good to solve the current issue, but still there seems to be a bug in the code as I pointed out in my last comment.

**Comment by nicolas-grekas** [comment]
@stefankleff indeed, I don't know why I ignored your comment but you were right, here is the PR with the fix #48971

**Comment by axi** [comment]
I confirm this fix my bug, thanks everyone
